design-proposals/klip-6-ksql-testing-tool.md

```diff
@@ -0,0 +1,186 @@
+
+Author: hjafarpour
+
+Release target: 5.3
+
+Status: In Discussion
+
+Discussion: _link to the design discussion PR_
+
+## tl;dr
+
+Provide a testing tool to evaluate the generated results from a set of KSQL statements without requiring the full fledge KSQL/Kafka environment.
+
+
+## Motivation and background
+
+This KLIP proposed a new testing tool for KSQL that can be used to evaluate the generated results for a set of KSQL statements. Currently, if users want to evaluate the result of KSQL statements, they should bring up a full fledged KSQL environment including Kafka cluster, KSQL cluster, KSQL client (e.g., KSQL CLI) and possibly Schema Registry if the test data is in Avro format.
+With the proposed testing tool, users can provide a set of KSQL statements along with specific configurations, data formats, the input data and the expected output data. The testing tool will execute the statements with the provided input data and assert that the output data is produced by the statements.
+The testing tool also enables users to easily share their test files without any dependency on the execution environment. Therefore, it will be much easier to collaborate and design streaming pipelines with KSQL!
+
+
+## What is in scope
+
+The testing tool will be a self sufficient command line tool where users can pass JSON files describing test settings and configuration and evaluate the generated results.
+
+## What is not in scope
+
+N/A
+
+## Value/Return
+
+The proposed tool provides a new way of testing KSQL statements. It also is self sufficient and does not need any running environment such as Kafka or KSQL cluster.
+
+
+## Public APIS
+
+### Test File
+
+The test file is a JSON file that describes the test settings and test data. The following is an example test file:
+
+```JSON
+{
+ "comments": [
+   "Any comments on this test file!"
+ ],
+ "tests": [
+   {
+     "name": "Sample test",
+     "statements": [
+       "CREATE STREAM TEST (ID bigint, NAME varchar, VALUE double) WITH (kafka_topic='test_topic', value_format='DELIMITED', key='ID');",
+       "CREATE STREAM S1 as SELECT name FROM test where id > 100;"
+     ],
+     "inputs": [
+       {"topic": "test_topic", "key": 0, "value": "0,zero,0.0"},
+       {"topic": "test_topic", "key": 100, "value": "100,100,0.0"},
+       {"topic": "test_topic", "key": 101, "value": "101,101,0.0"}
+     ],
+     "outputs": [
+       {"topic": "S1", "key": 101, "value": "101"}
+     ]
+   }
+ ]
+}
+```
+
+Each test file can include multiple tests. Each test can have the following attributes (these descriptions are the same as the ones described ([here](https://github.com/confluentinc/ksql/blob/master/ksql-engine/src/test/resources/query-validation-tests/README.md))):
+
+- _name_ (required): The name of the test case.
+description (optional): A description of what the test case is testing.
+- _format_ (optional): An array of multiple different formats to run the test case as, e.g. AVRO, JSON, DELIMITED.
```

Comment:
We should specify the default behavior if this optional `format` field is not present (in any documentation we publish for this testing tool, and also in the README linked to below as a place to go for more details).

Reply:
nit: I think its actually `formats` and its an array of formats to test with. Allowing for replacement of `${FORMAT}` within the statements.